src/net_processing.cpp

```diff
@@ -815,7 +815,12 @@ void PeerLogicValidation::ReattemptInitialBroadcast(CScheduler& scheduler) const
     std::set<uint256> unbroadcast_txids = m_mempool.GetUnbroadcastTxs();
 
     for (const uint256& txid : unbroadcast_txids) {
-        RelayTransaction(txid, *connman);
+        // Sanity check: all unbroadcast txns should exist in the mempool
+        if (!m_mempool.exists(txid)) {
```

Comment:
```suggestion
        if (m_mempool.exists(txid)) {
```

style-nit: (feel free to ignore)

When an `if` has an else branch, I generally recommend not using inversion in the condition, just to avoid double negation, which can make code harder to read. Obviously not an issue here, so feel free to ignore.

However, while this fights the symptom (unbroadcast txid exists), it doesn't prevent it from happening. Also, `RelayTransaction` is hopefully already a noop if the txid does not exist.

What do you think about adding this check to the mempool member function that adds the unbroadcast transactions to the set? This way, the error will be caught as early as possible.

Reply:
Good point about inversion on an if+else, I'll rearrange to make it more clear.

Yeah, I think it'd make sense to add this check to `AddUnbroadcastTx` ([here](https://github.com/bitcoin/bitcoin/blob/7bcc42b4035b878719d13201286e322989b415c5/src/txmempool.h#L705)).
Apart from making sure we only add txids that exist in the mempool to unbroadcast, I think this sanity check helps if a tx was removed from the mempool but somehow stayed in the unbroadcast set - which shouldn't happen, but just in case?

Reply:
Jup, better to do the check in two places than in one. This needs a mempool lookup, but the cost doesn't seem too much, as this is only done every 10 minutes or when a new transaction is added to the mempool anyway.